File: windows/src/Pomodoro.Core/IntegrationStatus.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Pomodoro.Core;

public enum SetupKind
{
    Ready,
    Blocked,
    Missing
}

public readonly record struct SetupState(SetupKind Kind, string Detail)
{
    public bool IsReady => Kind == SetupKind.Ready;

    public static SetupState Ready() => new(SetupKind.Ready, "Ready");
    public static SetupState Blocked(string detail) => new(SetupKind.Blocked, detail);
    public static SetupState Missing(string detail) => new(SetupKind.Missing, detail);
}

/// <summary>
/// What each integration still needs before it can run. The credential files
/// are inspected up front so setup can be guided rather than failing deep
/// inside an API call.
/// </summary>
public sealed class IntegrationStatus
{
    public SetupState GoogleClient { get; private set; } = SetupState.Ready();
    public SetupState GoogleToken { get; private set; } = SetupState.Ready();
    public SetupState Whistler { get; private set; } = SetupState.Ready();

    /// <summary>Non-secret summary of the Whistler settings, for display only.</summary>
    public IReadOnlyList<(string Label, string Value)> WhistlerSummary { get; private set; } = [];

    public bool GoogleReady => GoogleClient.IsReady && GoogleToken.IsReady;
    public bool WhistlerReady => GoogleReady && Whistler.IsReady;

    public event Action? Changed;

    public IntegrationStatus() => Refresh();

    public void Refresh()
    {
        // Checks the bundled copy as well as the per-user one: a release ships
        // a client beside the executable, and only looking in the data
        // directory made the app report it missing.
        var client = DataPaths.ExistingGoogleClient();
        GoogleClient = client is null
            ? SetupState.Missing("No OAuth client installed")
            : ValidateClient(client);

        GoogleToken = File.Exists(DataPaths.GoogleToken)
            ? SetupState.Ready()
            : SetupState.Missing("Not authorized yet");

        // Settings are shown; secrets are only ever reported present or absent,
        // and are never read into the UI at all.
        var settings = WhistlerConfig.ReadSettings();
        var summary = new List<(string, string)>
        {
            ("Server", settings.ApiUrl),
            ("Calendar", settings.CalendarId),
            ("Model", settings.Model)
        };
        if (settings.Email.Length > 0) summary.Add(("Account", settings.Email));
        WhistlerSummary = summary;

        var missing = new List<string>();
        if (!SecretStore.Has(SecretStore.OpenRouterKey) &&
            string.IsNullOrEmpty(Environment.GetEnvironmentVariable("OPENROUTER_API_KEY")))
        {
            missing.Add("OpenRouter API key");
        }
        if (!SecretStore.Has(SecretStore.WhistlerSession) &&
            !SecretStore.Has(WhistlerConfig.LegacyPassword))
        {
            missing.Add("Whistler sign-in");
        }
        Whistler = missing.Count switch
        {
            0 => SetupState.Ready(),
            2 => SetupState.Missing("Not configured"),
            _ => SetupState.Blocked("Missing " + string.Join(" and ", missing))
        };

        Changed?.Invoke();
    }

    /// <summary>A Google "Desktop app" client serializes with an installed or web section.</summary>
    private static SetupState ValidateClient(string path)
    {
        JsonNode? parsed;
        try
        {
            parsed = JsonNode.Parse(File.ReadAllText(path));
        }
        catch (Exception error) when (error is JsonException or IOException)
        {
            return SetupState.Blocked("Client file is not valid JSON");
        }
        if (parsed is not JsonObject obj) return SetupState.Blocked("Client file is not valid JSON");
        var client = (obj["installed"] ?? obj["web"]) as JsonObject;
        if (client is null) return SetupState.Blocked("Client file has no installed/web section");
        var id = client["client_id"]?.GetValue<string>();
        return string.IsNullOrEmpty(id)
            ? SetupState.Blocked("Client file has no client_id")
            : SetupState.Ready();
    }

    public static Dictionary<string, string> ReadEnv(string path)
    {
        var values = new Dictionary<string, string>(StringComparer.Ordinal);
        var text = AtomicFile.Read(path);
        if (text is null) return values;

        foreach (var line in text.Split('\n'))
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0 || trimmed[0] == '#') continue;
            var separator = trimmed.IndexOf('=');
            if (separator <= 0) continue;
            var key = trimmed[..separator].Trim();
            var value = trimmed[(separator + 1)..].Trim();
            if (value.Length >= 2 && value[0] == value[^1] && (value[0] == '"' || value[0] == '\''))
            {
                value = value[1..^1];
            }
            values[key] = value;
        }
        return values;
    }

    /// <summary>
    /// Installs a client JSON downloaded from the Google Cloud console.
    /// Returns null on success, or a message to show the user.
    /// </summary>
    public string? InstallGoogleClient(string source)
    {
        string raw;
        try { raw = File.ReadAllText(source); }
        catch (IOException) { return $"Could not read {Path.GetFileName(source)}."; }

        JsonNode? parsed;
        try { parsed = JsonNode.Parse(raw); }
        catch (JsonException) { return "That file is not valid JSON."; }

        var client = (parsed as JsonObject) is { } obj ? (obj["installed"] ?? obj["web"]) as JsonObject : null;
        if (client is null || string.IsNullOrEmpty(client["client_id"]?.GetValue<string>()))
        {
            return "That file is not a Google OAuth client. Download the JSON for a Desktop app client.";
        }

        DataPaths.EnsureDirectory();
        try { File.WriteAllText(DataPaths.GoogleClient, raw); }
        catch (IOException error) { return "Could not install the client file: " + error.Message; }

        Refresh();
        return null;
    }
}

File: macos/Sources/Pomodoro/Core/IntegrationStatus.swift
import Foundation
import AppKit
import Combine

/// What each integration still needs before it can run.
///
/// The bridges fail with a bare `FileNotFoundError` when a credential file is
/// absent, which reads as a crash rather than as "you have not set this up
/// yet". The app inspects the same files up front so setup can be guided
/// instead of guessed at.
enum SetupState: Equatable {
    case ready
    case blocked(String)
    case missing(String)

    var isReady: Bool { self == .ready }

    var symbol: String {
        switch self {
        case .ready: return "checkmark.circle.fill"
        case .blocked: return "exclamationmark.triangle.fill"
        case .missing: return "circle.dashed"
        }
    }

    var detail: String {
        switch self {
        case .ready: return "Ready"
        case .blocked(let message), .missing(let message): return message
        }
    }
}

@MainActor
final class IntegrationStatus: ObservableObject {
    @Published private(set) var python: SetupState = .ready
    @Published private(set) var googleClient: SetupState = .ready
    @Published private(set) var googleToken: SetupState = .ready
    @Published private(set) var whistler: SetupState = .ready
    /// Non-secret summary of the Whistler settings, for display only.
    @Published private(set) var whistlerSummary: [(String, String)] = []

    private var watchers: [FileWatcher] = []

    init() {
        refresh()
        watchers = [DataPaths.googleClient, DataPaths.googleToken, DataPaths.whistlerAccount].map { url in
            FileWatcher(url: url) { [weak self] in self?.refresh() }
        }
    }

    var googleReady: Bool { googleClient.isReady && googleToken.isReady }
    var whistlerReady: Bool { googleReady && whistler.isReady }

    func refresh() {
        let fm = FileManager.default

        python = Bridge.isAvailable
            ? .ready
            : .blocked("No python3 found. Set POMODORO_PYTHON to an interpreter path.")

        if let url = DataPaths.existingGoogleClient() {
            googleClient = Self.validateClient(at: url)
        } else {
            googleClient = .missing("No OAuth client installed")
        }

        googleToken = fm.fileExists(atPath: DataPaths.googleToken.path)
            ? .ready
            : .missing("Not authorized yet")

        // Settings are shown; secrets are only ever reported present or
        // absent, and are never read into the UI at all.
        let settings = WhistlerConfig.readSettings()
        var summary: [(String, String)] = [
            ("Server", settings.apiUrl),
            ("Calendar", settings.calendarId),
            ("Model", settings.model)
        ]
        if !settings.email.isEmpty { summary.append(("Account", settings.email)) }
        whistlerSummary = summary

        var missing: [String] = []
        if !SecretStore.has(SecretStore.openRouterKey)
            && (ProcessInfo.processInfo.environment["OPENROUTER_API_KEY"] ?? "").isEmpty {
            missing.append("OpenRouter API key")
        }
        if !SecretStore.has(SecretStore.whistlerSession)
            && !SecretStore.has(WhistlerConfig.legacyPassword) {
            missing.append("Whistler sign-in")
        }
        switch missing.count {
        case 0: whistler = .ready
        case 2: whistler = .missing("Not configured")
        default: whistler = .blocked("Missing \(missing.joined(separator: " and "))")
        }
    }

    /// The bridges accept a Google "Desktop app" client, which serializes with
    /// either an `installed` or a `web` section.
    private static func validateClient(at url: URL) -> SetupState {
        guard let data = try? Data(contentsOf: url),
              let parsed = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return .blocked("Client file is not valid JSON")
        }
        guard let client = (parsed["installed"] ?? parsed["web"]) as? [String: Any] else {
            return .blocked("Client file has no installed/web section")
        }
        guard let id = client["client_id"] as? String, !id.isEmpty else {
            return .blocked("Client file has no client_id")
        }
        return .ready
    }

    private static func readEnv(_ url: URL) -> [String: String] {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return [:] }
        var values: [String: String] = [:]
        for line in text.split(separator: "\n") {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("#"),
                  let separator = trimmed.firstIndex(of: "=") else { continue }
            let key = String(trimmed[trimmed.startIndex..<separator]).trimmingCharacters(in: .whitespaces)
            var value = String(trimmed[trimmed.index(after: separator)...]).trimmingCharacters(in: .whitespaces)
            if value.count >= 2, let first = value.first, let last = value.last,
               first == last, first == "\"" || first == "'" {
                value = String(value.dropFirst().dropLast())
            }
            values[key] = value
        }
        return values
    }

    /// Installs a client JSON the user downloaded from the Google Cloud
    /// console. Copying it into the data directory is exactly what the bridges
    /// look for, and is the step the old flow silently assumed had happened.
    /// Returns nil on success, or a message to show the user.
    func installGoogleClient(from source: URL) -> String? {
        guard let data = try? Data(contentsOf: source) else {
            return "Could not read \(source.lastPathComponent)."
        }
        guard let parsed = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let client = (parsed["installed"] ?? parsed["web"]) as? [String: Any],
              let id = client["client_id"] as? String, !id.isEmpty else {
            return "That file is not a Google OAuth client. Download the JSON for a Desktop app client."
        }
        DataPaths.ensureDirectory()
        do {
            if FileManager.default.fileExists(atPath: DataPaths.googleClient.path) {
                try FileManager.default.removeItem(at: DataPaths.googleClient)
            }
            try data.write(to: DataPaths.googleClient, options: .atomic)
        } catch {
            return "Could not install the client file: \(error.localizedDescription)"
        }
        refresh()
        return nil
    }

    func removeGoogleAuthorization() {
        try? FileManager.default.removeItem(at: DataPaths.googleToken)
        refresh()
    }
}
